File: dp/2338-count-the-number-of-ideal-arrays.py
import math
import unittest

# https://leetcode.com/problems/count-the-number-of-ideal-arrays/

# python3 -m unittest dp/2338-count-the-number-of-ideal-arrays.py

class Solution(unittest.TestCase):
    def idealArrays(self, n: int, maxValue: int) -> int:
        combs = [0] * 15
        for k in range(1, 15):
            combs[k] = math.comb(n-1,k-1)
        memo = [[None]*15 for _ in range(maxValue+1)]
        def dp(num: int, k: int) -> int:
            if memo[num][k] != None:
                return memo[num][k]
            # num is in the last pos, combs[k] is the 
            # combination of n-1 previous positions
            # k is the number of items to be chosen
            res = combs[k]
            nxt = num * 2
            while nxt <= maxValue:
                res += dp(nxt, k+1)
                nxt += num
            memo[num][k] = res
            return res
        return sum(dp(i,1) for i in range(1, maxValue+1)) % (10**9 + 7)

    def test(self):
        for n, maxValue, expected in [
            (2, 5, 10),
            (5, 3, 11),
        ]:
            output = self.idealArrays(n, maxValue)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: dp/2338-count-the-number-of-ideal-arrays.go
package dp

// https://leetcode.com/problems/count-the-number-of-ideal-arrays/

func idealArrays(n int, maxValue int) int {
	const MOD, maxK int = 1e9 + 7, 14

	comb := make([][]int, n)
	for i := 0; i < n; i++ {
		comb[i] = make([]int, maxK+1)
	}

	comb[0][0] = 1
	for i := 1; i < n; i++ {
		comb[i][0] = 1
		for j := 1; j <= maxK && j <= i; j++ {
			comb[i][j] = (comb[i-1][j-1] + comb[i-1][j]) % MOD
		}
	}

	memo := make([][]*int, maxValue+1)
	for i := range memo {
		memo[i] = make([]*int, maxK+1)
	}

	var dp func(num int, k int) int

	dp = func(num int, k int) int {
		if k > maxK {
			return 0
		}

		if memo[num][k] != nil {
			return *memo[num][k]
		}
		// num is in the last pos, coms[k] is the
		// combination of n-1 previous positions
		// k is the number of items to be chosen
		res := comb[n-1][k-1]
		for mult := num * 2; mult <= maxValue; mult += num {
			res = (res + dp(mult, k+1)) % MOD
		}

		memo[num][k] = &res

		return res
	}
	total := 0

	for num := 1; num <= maxValue; num++ {
		total = (total + dp(num, 1)) % MOD
	}

	return total
}
